File: sensor_decoder/src/query_viz_tool.cpp
#include <iostream>

#include "std_msgs/Int32MultiArray.h"
#include "std_msgs/Header.h"
#include "sensor_msgs/Image.h"
#include "sensor_decoder/Data.h"
#include "sensor_decoder/Object.h"
#include "sensor_decoder/Label.h"

#include "ros/ros.h"
#include "ros/time.h"
#include "ros/package.h"

#include "cv_bridge/cv_bridge.h"
#include "opencv2/opencv.hpp"
#include <opencv2/core.hpp>
#include <opencv2/videoio.hpp>
#include <opencv2/highgui.hpp>


using namespace std;
using namespace cv;

typedef pair<double, double> pdd;


string zfill(int n){
    if(n==0) return "000000";
    int digit = log10(n) + 1;
    string rt = "";
    for(int i=0;i<6-digit;i++) rt += "0";
    return rt+to_string(n);
}

class QueryVisualizer{
    public:
    ros::NodeHandle nh_;
    ros::Publisher pub_;
    ros::Subscriber sub_data_;
    ros::Subscriber sub_label_;

    cv_bridge::CvImage bridge;
    
    std_msgs::Header header;

    string data_name;
    stringstream data_path;
    
    string img_path;
    string pcd_path;

    bool valid_data;
    bool valid_label;
    int seq;

    Mat bev_map;
    Mat img_raw;

    int height = 600;
    int width = 600;
    double resolution = 0.1;

    double minX = -10.0;
    double maxX = 20.0;
    double minY = -10.0;
    double maxY = 10.0;
    double minZ = 0.5;
    double maxZ = 2.0;

    
    QueryVisualizer(){
        data_name = "";
        data_path << ros::package::getPath("sensor_decoder") << "/data/" << data_name << "/";
        valid_data = false;
        valid_label = false;

        img_path = data_path.str() + "image_raw/";
        pcd_path = data_path.str() + "local_map/";

        sub_data_ = nh_.subscribe("/query_data", 1, &QueryVisualizer::callback_data, this);
        sub_label_ = nh_.subscribe("/query_label", 1, &QueryVisualizer::callback_label, this);
        pub_ = nh_.advertise<sensor_msgs::Image>("query_image", 10);
        // query();
    }

    cv::Point xy_to_pixel(pdd pos){
        return cv::Point(width/2 + int(pos.first/resolution), height - int(pos.second/resolution));
    }

    void draw_object(const sensor_decoder::Object& obj){
        
    }

    void callback_data(const sensor_decoder::Data::ConstPtr& msg){
        data_name = msg->name;
        data_path << ros::package::getPath("sensor_decoder") << "/data" << data_name << "/";

        img_path = data_path.str() + "image_raw/";
        pcd_path = data_path.str() + "pcd/";

        string cur_image_raw_path = image_raw_path + zfill(seq) + ".png";
        img_raw = cv::imread(cur_image_raw_path);

        bev_map = Mat(height, width, CV_8UC3, Scalar(255,255,255));

        string pcd_file = pcd_path + zfill(seq) + ".pcd";

        pcl::PointCloud<pcl::PointXYZ>::Ptr cloud (new pcl::PointCloud<pcl::PointXYZ>);
            
        if (pcl::io::loadPCDFile<pcl::PointXYZ> (pcd_file, *cloud) == -1) {
            PCL_ERROR ("Couldn't read file test_pcd.pcd \n");
            return (-1);
        }

        for(const auto& p : *cloud) {
            double x = p.x;
            double y = p.y;
            double z = p.z;
            if(x < minX + EPS || x > maxX - EPS || y < minY + EPS || y > maxY - EPS || z < minZ || z > maxZ) continue;
            
            cv::Point pixel = xy_to_pixel({x, y});
            bev_map.at<cv::Vec3b>(pixel[1], pixel[0])[0] = 0;
            bev_map.at<cv::Vec3b>(pixel[1], pixel[0])[1] = 0;
            bev_map.at<cv::Vec3b>(pixel[1], pixel[0])[2] = 0;
        }


        valid = true;
    }

    void callback_label(const sensor_decoder::Label::ConstPtr& msg){
        

        return;
    }



    void query(){
        if(!valid_data || !valid_label) return;

        

        // Mat local_map = pc_map;

        
    }
};

int main(int argc, char **argv){
    ros::init(argc, argv, "query_viz_tool");
    QueryVisualizer q;
    ros::Rate loop_rate(10);
    while(ros::ok()){
        // q.query();
        // if(!q.valid) continue;
        // int seq = q.start + q.N % q.n_data;
        // string cur_image_raw_path = q.image_raw_path + zfill(seq) + ".png";
        // string cur_local_map_path = q.local_map_path + zfill(seq) + ".png";

        // Mat image_raw = cv::imread(cur_image_raw_path);
        // Mat local_map = Mat(img_height_, img_width_, CV_8UC3, Scalar(255,255,255));
        // Mat local_map = cv::imread(cur_local_map_path);
        // if(image_raw.size().height == 0 || local_map.size().height == 0) continue;
        // Mat out;

        // cv::resize(image_raw, image_raw, Size(300,300), 0, 0);
        // cv::resize(local_map, local_map, Size(300,300), 0, 0);

        // Mat arr[2] = {image_raw, local_map};
        // cv::hconcat(arr,2,out);
        // sensor_msgs::Image rt;
        // q.bridge = cv_bridge::CvImage(std_msgs::Header(), sensor_msgs::image_encodings::RGB8, out);
        // q.bridge.toImageMsg(rt);
        // q.pub_.publish(rt);
        // q.N++;
        // ros::spinOnce();
        // loop_rate.sleep();
    }
    ros::spin();
}

File: sensor_decoder/scripts/query_viz_tool.py
#!/usr/bin/python
import rospy
import rospkg

from sensor_decoder.msg import Data
from sensor_decoder.msg import Object
from sensor_decoder.msg import Label
from sensor_msgs.msg import Image

import numpy as np
import math
import cv2
from cv_bridge import CvBridge
import copy

# minX = -10.0
# maxX = 20.0
# minY = -10.0
# maxY = 10.0
# minZ = 0.5
# maxZ = 2.0
resolution = 0.1
bev_height = 600
bev_width = 600

"""
For bird eye view map
"""
minX = -20.0
maxX = 40.0
minY = -30.0
maxY = 30.0
minZ = 0.5
maxZ = 2.0

cx = (minX + maxX) / 2.0
cy = (minY + maxY) / 2.0

maxlogDensity = math.log(20)
"""
For local map
"""
font = cv2.FONT_HERSHEY_SIMPLEX
fontscale = 0.5
fontthickness = 2
fontline = cv2.LINE_AA

def get_pixel(x, y):
    px = int((x - maxX) / (minX - maxX) * bev_height)
    py = int((y - maxY) / (minY - maxY) * bev_width)

    return px, py

class QueryVizTool:
    def __init__(self):
        self.pub = rospy.Publisher("/query_image", Image, queue_size=10)
        self.seq = 0
        self.bin_path = ""
        self.bridge = CvBridge()
        self.valid_data = False
        self.valid_label = False

        self.sub_data = rospy.Subscriber("/query_data", Data, self.callback_data)
        self.sub_label = rospy.Subscriber("/query_label", Label, self.callback_label)

    def draw_object(self, objects):
        local_map = self.bev_map
        for i in range(len(objects)):
            cx = objects[i].x
            cy = objects[i].y
            l = objects[i].l
            w = objects[i].w
            theta = objects[i].theta
            box = []
            box.append(get_pixel(cx + l/2 * math.cos(theta) - w/2 * math.sin(theta), cy + l/2 * math.sin(theta) + w/2 * math.cos(theta)))
            box.append(get_pixel(cx - l/2 * math.cos(theta) - w/2 * math.sin(theta), cy - l/2 * math.sin(theta) + w/2 * math.cos(theta)))
            box.append(get_pixel(cx - l/2 * math.cos(theta) + w/2 * math.sin(theta), cy - l/2 * math.sin(theta) - w/2 * math.cos(theta)))
            box.append(get_pixel(cx + l/2 * math.cos(theta) + w/2 * math.sin(theta), cy + l/2 * math.sin(theta) - w/2 * math.cos(theta)))
            box.append(get_pixel(cx, cy))
            if objects[i].valid:
                cv2.line(local_map, (box[0][1], box[0][0]), (box[1][1], box[1][0]), (0, 0, 255), 2)
                cv2.line(local_map, (box[1][1], box[1][0]), (box[2][1], box[2][0]), (0, 0, 255), 2)
                cv2.line(local_map, (box[2][1], box[2][0]), (box[3][1], box[3][0]), (0, 0, 255), 2)
                cv2.line(local_map, (box[3][1], box[3][0]), (box[0][1], box[0][0]), (0, 255, 255), 2)
                cv2.putText(local_map, format(objects[i].id,"d"), (box[4][1], box[4][0]), font, fontscale, (255, 255, 255), fontthickness, fontline)
            else :
                if not self.only_valid :    
                    cv2.line(local_map, (box[0][1], box[0][0]), (box[1][1], box[1][0]), (255, 0, 0), 1)
                    cv2.line(local_map, (box[1][1], box[1][0]), (box[2][1], box[2][0]), (255, 0, 0), 1)
                    cv2.line(local_map, (box[2][1], box[2][0]), (box[3][1], box[3][0]), (255, 0, 0), 1)
                    cv2.line(local_map, (box[3][1], box[3][0]), (box[0][1], box[0][0]), (255, 255, 0), 1)
                    cv2.putText(local_map, format(objects[i].id,"d"), (box[4][1], box[4][0]), font, fontscale, (255, 255, 255), fontthickness, fontline)
  
        self.local_map = local_map

    def callback_data(self, msg):
        # generate bev map & image
        # build bev map
        self.data_name = msg.name
        self.seq = msg.seq

        self.data_path = rospkg.RosPack().get_path("sensor_decoder") + "/data/" + self.data_name + "/"
        self.img_path = self.data_path + "image_raw/"
        self.bin_path = self.data_path + "bin/"

        self.bin_file = self.bin_path + str(self.seq).zfill(6) + ".bin"
        self.img_file = self.img_path + str(self.seq).zfill(6) + ".png"
        self.img_raw = cv2.imread(self.img_file, cv2.COLOR_BGR2RGB)
        pcs = np.fromfile(self.bin_file, dtype=np.float32).reshape(-1, 4)
        x = pcs[:,0]
        y = pcs[:,1]
        z = pcs[:,2]
        intensity = pcs[:,3]
        
        indices = []
        for i in range(len(pcs)):
            if x[i] > minX and x[i] < maxX and y[i] > minY and y[i] < maxY and z[i] > minZ and z[i] < maxZ:
                indices.append(i)
        pcs = pcs[indices,:]
        x = x[indices]
        y = y[indices]
        z = z[indices]
        intensity = intensity[indices]
        n_points = len(intensity)
 
        
        maxlogDensity = math.log(20)

        intensity_layer = np.zeros([bev_height, bev_width], dtype=np.float)
        density_layer = np.zeros([bev_height, bev_width], dtype=np.float)
        height_layer = np.zeros([bev_height, bev_width], dtype=np.float)
        for i in range(n_points):
            px, py = get_pixel(x[i], y[i])
            if px < 0 or px >= bev_height or py < 0 or py >= bev_width:
                continue
            intensity_layer[px][py] = max(intensity_layer[px][py], intensity[i])
            density_layer[px][py] += 1
            height_layer[px][py] = max(height_layer[px][py], (z[i]-minZ)/ (maxZ-minZ))
        for i in range(bev_height):
            for j in range(bev_width):
                density_layer[px][py] = min(1.0, math.log(1 + density_layer[px][py]) / maxlogDensity)
        intensity_layer = intensity_layer * 255.0
        density_layer = density_layer * 255.0
        height_layer = height_layer * 255.0
        intensity_layer = np.expand_dims(intensity_layer.astype('uint8'), axis = 0)
        density_layer = np.expand_dims(density_layer.astype('uint8'), axis = 0)
        height_layer = np.expand_dims(height_layer.astype('uint8'), axis = 0)
        local_map = np.transpose(np.vstack((intensity_layer, density_layer, height_layer)), (1,2,0))
        local_map = cv2.resize(local_map, (bev_height, bev_width))

        self.bev_map = local_map

        self.valid_data = True

    def callback_label(self, msg):
        # update local map
        self.only_valid = msg.only_valid
        self.draw_object(msg.objects)
        self.valid_label = True


    def query(self):
        if not self.valid_data or not self.valid_label :
            return
        img = cv2.hconcat([cv2.resize(self.img_raw, dsize=(600,600)), self.local_map])
        self.pub.publish(self.bridge.cv2_to_imgmsg(img))
    



if __name__=='__main__':
    rospy.init_node("query_viz_tool", anonymous=True)
    q = QueryVizTool()
    rate = rospy.Rate(20)
    while not rospy.is_shutdown():
        q.query()
        rate.sleep()